### ISSUE TITLE
Epic: BGE-small sentence scorer and multi-task training pipeline

### ISSUE THREAD

**Issue by tartakovsky** [berelevant-ai/slopless#18]
# BGE-small sentence scorer and multi-task training pipeline

Parent issue tracking the neural scorer track for slopless. The scorer is a parallel research effort to the deterministic rule family — **no textlint rule depends on it until it passes evaluation**, and the runtime stays opt-in (likely a separate package eventually).

Plan: [`.plans/2026-05-13-164606-bge-small-sentence-scorer-prototype.md`](https://github.com/agent-quality-controls/slopless/blob/main/.plans/2026-05-13-164606-bge-small-sentence-scorer-prototype.md).

## Architecture

```
sentence
  -> bge-small encoder (BAAI/bge-small-en-v1.5)
  -> pooled sentence vector
  -> 4 separate heads
  -> 4 separate scores
```

Heads:

| head | output | label source |
|---|---|---|
| `specificity_head` | regression 0..1 | Domain-Agnostic Sentence Specificity (news binary + Twitter/Yelp/Movies real-valued) |
| `vagueness_head` | regression or ordinal 0..1 | Lebanoff/Liu vagueness dataset (sentence + phrase) |
| `informativeness_head` | regression 0..1 | SQUINKY corpus (7,032 sentences) |
| `boilerplate_head` | binary / probability | custom-labeled, repeated-segment method |

Grounding is **not** a neural head. It stays as deterministic Lancaster + Brysbaert lookups.

## Why BGE-small

Measured on this machine:
- BGE-small ONNX: 127.8 MB, 330–380 sentences/sec at batch 8–64
- Voyage-4-nano ONNX: 240.7 MB, 38–44 sentences/sec

BGE-small is standard BERT-style; Voyage uses a custom Qwen3 bidirectional setup that complicates export and tuning. Voyage stays in scope as a later comparison if BGE-small head-only quality fails.

## Training phases

1. **Phase 1 — frozen encoder, head-only.** Train each head against existing labeled data with the encoder fixed. Heads can be LightGBM, logistic regression, ridge, or small MLP. Goal: establish per-axis metrics before any fine-tuning.
2. **Phase 2 — multi-task fine-tune.** Update encoder + heads jointly. Low learning rate on encoder. Per-axis masked loss: a sentence only contributes to a head's loss if it has a label for that head.
3. **Phase 3 — runtime export.** ONNX (fp32 baseline), then int8 quantization, then benchmark output drift.

## Tracks

This epic decomposes into four tracks, each with its own sub-issues:

- **Encoder & plumbing** — BGE-small integration, training pipeline shell, runtime export
- **Heads** — one sub-issue per axis (specificity, vagueness, informativeness, boilerplate)
- **Datasets** — prep per axis, including the LLM-labeling pipeline that enriches each existing dataset
- **Labeling pipeline** — Structured-Outputs prompts, calibration loop, active-learning loop

## Cross-cutting constraints (from plan)

- Local-only runtime. No network calls at inference time.
- Artifacts in the low hundreds of MB or less.
- Outputs are **per-axis scores**, never a single combined "slop" label.
- `rationale_short` from LLM labels is for audit only. Do not train on rationales first.
- Synthetic data only for contrast pairs and stress tests, never as the primary positive set.

## Done

Epic closes when:
- All 13 sub-issues are closed.
- The scorer passes the evaluation thresholds in the plan's "Evaluation" section.
- A decision is made on whether to ship the scorer inside `slopless` or as a separate package.
